# Lab Assignment 2: How to Load CSV, ASCII, and other data into Python
## DS 6001: Practice and Application of Data Science
### Drew Haynes (rbc6wr)

### Instructions
Please answer the following questions as completely as possible using text, code, and the results of code as needed. Format your answers in a Jupyter notebook. To receive full credit, make sure you address every part of the problem, and make sure your document is formatted in a clean and professional way.

There are 11 data files attached to this lab assignment, with different extensions. First, download all of these data files, and save them in the same folder on your local machine. Your task in the following questions is to load each file into Python correctly, so that you can begin the process of data cleaning. If the variable names are included in the file, use those names to name the columns. If the variable names are not included, use these names in order:

In [1]:
column_names = ["Country", "Happiness score", "Whisker-high", "Whisker-low", 
  "Dystopia (1.92) + residual", "Explained by: GDP per capita", 
  "Explained by: Social support", "Explained by: Healthy life expectancy", 
  "Explained by: Freedom to make life choices", "Explained by: Generosity", 
  "Explained by: Perceptions of corruption" ]

If you loaded the data correctly, it will look like `data_clean.csv`, which is also attached to this lab.

## Problem 0
Import the libraries you will need. Then write code to change the working directory to the folder in which you saved the data files, run the code displayed above to create the `column_names` list, load `data_clean.csv`, and display the output of the `.info()` method of `data_clean`. (1 point)

## Answer 0

In [2]:
import numpy as np
import pandas as pd
import os
import csv

In [3]:
os.chdir('.\lab data')

In [4]:
df_clean = pd.read_csv('data_clean.csv')
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     156 non-null    object 
 1   Happiness score                             156 non-null    float64
 2   Whisker-high                                156 non-null    float64
 3   Whisker-low                                 156 non-null    float64
 4   Dystopia (1.92) + residual                  156 non-null    float64
 5   Explained by: GDP per capita                156 non-null    float64
 6   Explained by: Social support                156 non-null    float64
 7   Explained by: Healthy life expectancy       156 non-null    float64
 8   Explained by: Freedom to make life choices  156 non-null    float64
 9   Explained by: Generosity                    156 non-null    float64
 10  Explained by: 

## Problem 1
Load `data1.csv`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

## Answer 1

In [5]:
df1 = pd.read_csv('data1.csv')

In [6]:
df1.info()
df1.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158 entries, 0 to 157
Data columns (total 11 columns):
 #   Column                                                                                           Non-Null Count  Dtype 
---  ------                                                                                           --------------  ----- 
 0   Source: The World Happiness Report (2018), The Sustainable Development Solutions Network (SDSN)  158 non-null    object
 1   Unnamed: 1                                                                                       157 non-null    object
 2   Unnamed: 2                                                                                       157 non-null    object
 3   Unnamed: 3                                                                                       157 non-null    object
 4   Unnamed: 4                                                                                       157 non-null    object
 5   Unnamed: 5     

Source: The World Happiness Report (2018), The Sustainable Development Solutions Network (SDSN)  \
count                                                 158                                                
unique                                                158                                                
top             URL: http://worldhappiness.report/ed/2018                                                
freq                                                    1                                                

       Unnamed: 1    Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
count         157           157        157        157        157        157   
unique        156           157        152        148        148        147   
top         5.483  Whisker-high      5.035      2.319      1.229      1.161   
freq            2             1          2          2          2          2   

       Unnamed: 7 Unnamed: 8 Unnamed: 9 Unnamed: 10  
count         157        157        157         157  
unique        143        137        123         112  
top         0.737      0.531      0.092       0.082  
freq            2          3          3           5

In [7]:
df1.head(10)

Source: The World Happiness Report (2018), The Sustainable Development Solutions Network (SDSN)  \
0          URL: http://worldhappiness.report/ed/2018                                                
1                                            Country                                                
2                                            Finland                                                
3                                             Norway                                                
4                                            Denmark                                                
5                                            Iceland                                                
6                                        Switzerland                                                
7                                        Netherlands                                                
8                                             Canada                                                
9                                        New Zealand                                                

        Unnamed: 1    Unnamed: 2   Unnamed: 3                  Unnamed: 4  \
0              NaN           NaN          NaN                         NaN   
1  Happiness score  Whisker-high  Whisker-low  Dystopia (1.92) + residual   
2            7.632         7.695        7.569                       2.595   
3            7.594         7.657        7.530                       2.383   
4            7.555         7.623        7.487                       2.370   
5            7.495         7.593        7.398                       2.426   
6            7.487         7.570        7.405                       2.320   
7            7.441         7.498        7.384                       2.448   
8            7.328         7.421        7.236                       2.306   
9            7.324         7.401        7.247                       2.156   

                     Unnamed: 5                    Unnamed: 6  \
0                           NaN                           NaN   
1  Explained by: GDP per capita  Explained by: Social support   
2                         1.305                         1.592   
3                         1.456                         1.582   
4                         1.351                         1.590   
5                         1.343                         1.644   
6                         1.420                         1.549   
7                         1.361                         1.488   
8                         1.330                         1.532   
9                         1.268                         1.601   

                              Unnamed: 7  \
0                                    NaN   
1  Explained by: Healthy life expectancy   
2                                  0.874   
3                                  0.861   
4                                  0.868   
5                                  0.914   
6                                  0.927   
7                                  0.878   
8                                  0.896   
9                                  0.876   

                                   Unnamed: 8                Unnamed: 9  \
0                                         NaN                       NaN   
1  Explained by: Freedom to make life choices  Explained by: Generosity   
2                                       0.681                     0.192   
3                                       0.686                     0.286   
4                                       0.683                     0.284   
5                                       0.677                     0.353   
6                                       0.660                     0.256   
7                                       0.638                     0.333   
8                                       0.653                     0.321   
9                                       0.669                     0.365   

               

In [8]:
df1 = pd.read_csv('data1.csv', header = 2)
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     156 non-null    object 
 1   Happiness score                             156 non-null    float64
 2   Whisker-high                                156 non-null    float64
 3   Whisker-low                                 156 non-null    float64
 4   Dystopia (1.92) + residual                  156 non-null    float64
 5   Explained by: GDP per capita                156 non-null    float64
 6   Explained by: Social support                156 non-null    float64
 7   Explained by: Healthy life expectancy       156 non-null    float64
 8   Explained by: Freedom to make life choices  156 non-null    float64
 9   Explained by: Generosity                    156 non-null    float64
 10  Explained by: 

In [9]:
df1.head(10)

Country  Happiness score  Whisker-high  Whisker-low  \
0      Finland            7.632         7.695        7.569   
1       Norway            7.594         7.657        7.530   
2      Denmark            7.555         7.623        7.487   
3      Iceland            7.495         7.593        7.398   
4  Switzerland            7.487         7.570        7.405   
5  Netherlands            7.441         7.498        7.384   
6       Canada            7.328         7.421        7.236   
7  New Zealand            7.324         7.401        7.247   
8       Sweden            7.314         7.388        7.241   
9    Australia            7.272         7.358        7.186   

   Dystopia (1.92) + residual  Explained by: GDP per capita  \
0                       2.595                         1.305   
1                       2.383                         1.456   
2                       2.370                         1.351   
3                       2.426                         1.343   
4                       2.320                         1.420   
5                       2.448                         1.361   
6                       2.306                         1.330   
7                       2.156                         1.268   
8                       2.217                         1.355   
9                       2.139                         1.340   

   Explained by: Social support  Explained by: Healthy life expectancy  \
0                         1.592                                  0.874   
1                         1.582                                  0.861   
2                         1.590                                  0.868   
3                         1.644                                  0.914   
4                         1.549                                  0.927   
5                         1.488                                  0.878   
6                         1.532                                  0.896   
7                         1.601                                  0.876   
8                         1.501                                  0.913   
9                         1.573                                  0.910   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                       0.681                     0.192   
1                                       0.686                     0.286   
2                                       0.683                     0.284   
3                                       0.677                     0.353   
4                                       0.660                     0.256   
5                                       0.638                     0.333   
6                                       0.653                     0.321   
7                                       0.669                     0.365   
8                                       0.659                     0.285   
9                                       0.647                     0.361   

   Explained by: Perceptions of corruption  
0                                    0.393  
1                                    0.340  
2                                    0.408  
3                                    0.138  
4                                    0.357  
5                                    0.295  
6                                    0.291  
7                                    0.389  
8                                    0.383  
9                                    0.302

So it looks like the first row contains the source URL and headers belong on the second. By passing `header = 2` as a parameter in the `read_csv` function, it looks like things now load correctly with a constant Dtype with columns and no null rows.

## Problem 2
Load `data2.txt`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

## Answer 2

In [10]:
df2 = pd.read_csv('data2.txt')
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 161 entries, 0 to 160
Data columns (total 11 columns):
 #   Column                                                                                           Non-Null Count  Dtype 
---  ------                                                                                           --------------  ----- 
 0   Source: The World Happiness Report (2018), The Sustainable Development Solutions Network (SDSN)  161 non-null    object
 1   Unnamed: 1                                                                                       157 non-null    object
 2   Unnamed: 2                                                                                       157 non-null    object
 3   Unnamed: 3                                                                                       157 non-null    object
 4   Unnamed: 4                                                                                       157 non-null    object
 5   Unnamed: 5     

In [11]:
df2.head(10)

Source: The World Happiness Report (2018), The Sustainable Development Solutions Network (SDSN)  \
0          URL: http://worldhappiness.report/ed/2018                                                
1                                            Country                                                
2  /The following countries comprise the "very ha...                                                
3                                            Finland                                                
4                                             Norway                                                
5                                            Denmark                                                
6                                            Iceland                                                
7                                        Switzerland                                                
8                                        Netherlands                                                
9                                             Canada                                                

        Unnamed: 1    Unnamed: 2   Unnamed: 3                  Unnamed: 4  \
0              NaN           NaN          NaN                         NaN   
1  Happiness score  Whisker-high  Whisker-low  Dystopia (1.92) + residual   
2              NaN           NaN          NaN                         NaN   
3            7.632         7.695        7.569                       2.595   
4            7.594         7.657        7.530                       2.383   
5            7.555         7.623        7.487                       2.370   
6            7.495         7.593        7.398                       2.426   
7            7.487         7.570        7.405                       2.320   
8            7.441         7.498        7.384                       2.448   
9            7.328         7.421        7.236                       2.306   

                     Unnamed: 5                    Unnamed: 6  \
0                           NaN                           NaN   
1  Explained by: GDP per capita  Explained by: Social support   
2                           NaN                           NaN   
3                         1.305                         1.592   
4                         1.456                         1.582   
5                         1.351                         1.590   
6                         1.343                         1.644   
7                         1.420                         1.549   
8                         1.361                         1.488   
9                         1.330                         1.532   

                              Unnamed: 7  \
0                                    NaN   
1  Explained by: Healthy life expectancy   
2                                    NaN   
3                                  0.874   
4                                  0.861   
5                                  0.868   
6                                  0.914   
7                                  0.927   
8                                  0.878   
9                                  0.896   

                                   Unnamed: 8                Unnamed: 9  \
0                                         NaN                       NaN   
1  Explained by: Freedom to make life choices  Explained by: Generosity   
2                                         NaN                       NaN   
3                                       0.681                     0.192   
4                                       0.686                     0.286   
5                                       0.683                     0.284   
6                                       0.677                     0.353   
7                                       0.660                     0.256   
8                                       0.638                     0.333   
9                                       0.653                     0.321   

               

In [12]:
df2 = pd.read_csv('data2.txt', header = 2)
df2 = df2[df2.Country.str.contains("/")==False]
df2.info()
df2.describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 156 entries, 1 to 158
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     156 non-null    object 
 1   Happiness score                             156 non-null    float64
 2   Whisker-high                                156 non-null    float64
 3   Whisker-low                                 156 non-null    float64
 4   Dystopia (1.92) + residual                  156 non-null    float64
 5   Explained by: GDP per capita                156 non-null    float64
 6   Explained by: Social support                156 non-null    float64
 7   Explained by: Healthy life expectancy       156 non-null    float64
 8   Explained by: Freedom to make life choices  156 non-null    float64
 9   Explained by: Generosity                    156 non-null    float64
 10  Explained by: 

Happiness score  Whisker-high  Whisker-low  Dystopia (1.92) + residual  \
count       156.000000    156.000000   156.000000                  156.000000   
mean          5.375878      5.478654     5.273090                    1.922891   
std           1.119507      1.107522     1.132544                    0.513760   
min           2.905000      3.074000     2.735000                    0.292000   
25%           4.453750      4.589750     4.344750                    1.654250   
50%           5.378000      5.477500     5.284500                    1.908500   
75%           6.168500      6.260000     6.051250                    2.270250   
max           7.632000      7.695000     7.569000                    2.961000   

       Explained by: GDP per capita  Explained by: Social support  \
count                    156.000000                    156.000000   
mean                       0.887417                      1.216571   
std                        0.382573                      0.300380   
min                        0.000000                      0.000000   
25%                        0.616250                      1.076750   
50%                        0.949500                      1.261500   
75%                        1.197750                      1.463000   
max                        1.649000                      1.644000   

       Explained by: Healthy life expectancy  \
count                             156.000000   
mean                                0.598026   
std                                 0.247925   
min                                 0.000000   
25%                                 0.422250   
50%                                 0.644000   
75%                                 0.777250   
max                                 1.030000   

       Explained by: Freedom to make life choices  Explained by: Generosity  \
count                                  156.000000                156.000000   
mean                                     0.456981                  0.181571   
std                                      0.162750                  0.098805   
min                                      0.000000                  0.000000   
25%                                      0.358250                  0.109500   
50%                                      0.494000                  0.174000   
75%                                      0.580000                  0.242250   
max                                      0.724000                  0.598000   

       Explained by: Perceptions of corruption  
count                               156.000000  
mean                                  0.112474  
std                                   0.096362  
min                                   0.000000  
25%                                   0.051000  
50%                                   0.082000  
75%                                   0.139000  
max                                   0.457000

This dataset had a couple comment lines indiciated by "/" throughout the dataset that needed filtering out. I noticed there was an additional comment line due to the non-null count not matching.

## Problem 3
Load `data3.txt`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

## Answer 3

In [13]:
df3 = pd.read_csv('data3.txt')
df3.head(10)

Source: The World Happiness Report (2018), The Sustainable Development Solutions Network (SDSN)\t\t\t\t\t\t\t\t\t\t
0  URL: http://worldhappiness.report/ed/2018\t\t\...                                                                 
1  Country\tHappiness score\tWhisker-high\tWhiske...                                                                 
2  Finland\t7.632\t7.695\t7.569\t2.595\t1.305\t1....                                                                 
3  Norway\t7.594\t7.657\t7.53\t2.383\t1.456\t1.58...                                                                 
4  Denmark\t7.555\t7.623\t7.487\t2.37\t1.351\t1.5...                                                                 
5  Iceland\t7.495\t7.593\t7.398\t2.426\t1.343\t1....                                                                 
6  Switzerland\t7.487\t7.57\t7.405\t2.32\t1.42\t1...                                                                 
7  Netherlands\t7.441\t7.498\t7.384\t2.448\t1.361...                                                                 
8  Canada\t7.328\t7.421\t7.236\t2.306\t1.33\t1.53...                                                                 
9  New Zealand\t7.324\t7.401\t7.247\t2.156\t1.268...

In [14]:
df3 = pd.read_csv('data3.txt', sep='\t', header = 2)
df3.info()
df3.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     156 non-null    object 
 1   Happiness score                             156 non-null    float64
 2   Whisker-high                                156 non-null    float64
 3   Whisker-low                                 156 non-null    float64
 4   Dystopia (1.92) + residual                  156 non-null    float64
 5   Explained by: GDP per capita                156 non-null    float64
 6   Explained by: Social support                156 non-null    float64
 7   Explained by: Healthy life expectancy       156 non-null    float64
 8   Explained by: Freedom to make life choices  156 non-null    float64
 9   Explained by: Generosity                    156 non-null    float64
 10  Explained by: 

Country  Happiness score  Whisker-high  Whisker-low  \
0      Finland            7.632         7.695        7.569   
1       Norway            7.594         7.657        7.530   
2      Denmark            7.555         7.623        7.487   
3      Iceland            7.495         7.593        7.398   
4  Switzerland            7.487         7.570        7.405   
5  Netherlands            7.441         7.498        7.384   
6       Canada            7.328         7.421        7.236   
7  New Zealand            7.324         7.401        7.247   
8       Sweden            7.314         7.388        7.241   
9    Australia            7.272         7.358        7.186   

   Dystopia (1.92) + residual  Explained by: GDP per capita  \
0                       2.595                         1.305   
1                       2.383                         1.456   
2                       2.370                         1.351   
3                       2.426                         1.343   
4                       2.320                         1.420   
5                       2.448                         1.361   
6                       2.306                         1.330   
7                       2.156                         1.268   
8                       2.217                         1.355   
9                       2.139                         1.340   

   Explained by: Social support  Explained by: Healthy life expectancy  \
0                         1.592                                  0.874   
1                         1.582                                  0.861   
2                         1.590                                  0.868   
3                         1.644                                  0.914   
4                         1.549                                  0.927   
5                         1.488                                  0.878   
6                         1.532                                  0.896   
7                         1.601                                  0.876   
8                         1.501                                  0.913   
9                         1.573                                  0.910   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                       0.681                     0.192   
1                                       0.686                     0.286   
2                                       0.683                     0.284   
3                                       0.677                     0.353   
4                                       0.660                     0.256   
5                                       0.638                     0.333   
6                                       0.653                     0.321   
7                                       0.669                     0.365   
8                                       0.659                     0.285   
9                                       0.647                     0.361   

   Explained by: Perceptions of corruption  
0                                    0.393  
1                                    0.340  
2                                    0.408  
3                                    0.138  
4                                    0.357  
5                                    0.295  
6                                    0.291  
7                                    0.389  
8                                    0.383  
9                                    0.302

In [15]:
df3.describe()

Happiness score  Whisker-high  Whisker-low  Dystopia (1.92) + residual  \
count       156.000000    156.000000   156.000000                  156.000000   
mean          5.375878      5.478654     5.273090                    1.922891   
std           1.119507      1.107522     1.132544                    0.513760   
min           2.905000      3.074000     2.735000                    0.292000   
25%           4.453750      4.589750     4.344750                    1.654250   
50%           5.378000      5.477500     5.284500                    1.908500   
75%           6.168500      6.260000     6.051250                    2.270250   
max           7.632000      7.695000     7.569000                    2.961000   

       Explained by: GDP per capita  Explained by: Social support  \
count                    156.000000                    156.000000   
mean                       0.887417                      1.216571   
std                        0.382573                      0.300380   
min                        0.000000                      0.000000   
25%                        0.616250                      1.076750   
50%                        0.949500                      1.261500   
75%                        1.197750                      1.463000   
max                        1.649000                      1.644000   

       Explained by: Healthy life expectancy  \
count                             156.000000   
mean                                0.598026   
std                                 0.247925   
min                                 0.000000   
25%                                 0.422250   
50%                                 0.644000   
75%                                 0.777250   
max                                 1.030000   

       Explained by: Freedom to make life choices  Explained by: Generosity  \
count                                  156.000000                156.000000   
mean                                     0.456981                  0.181571   
std                                      0.162750                  0.098805   
min                                      0.000000                  0.000000   
25%                                      0.358250                  0.109500   
50%                                      0.494000                  0.174000   
75%                                      0.580000                  0.242250   
max                                      0.724000                  0.598000   

       Explained by: Perceptions of corruption  
count                               156.000000  
mean                                  0.112474  
std                                   0.096362  
min                                   0.000000  
25%                                   0.051000  
50%                                   0.082000  
75%                                   0.139000  
max                                   0.457000

This dataset had a different delimiter, which was addressed with passing the parameter `sep='\t'`.

## Problem 4
Load `data4.txt`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

## Answer 4

In [16]:
df4 = pd.read_csv('data4.txt')
df4.info()
df4.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155 entries, 0 to 154
Data columns (total 1 columns):
 #   Column                                                               Non-Null Count  Dtype 
---  ------                                                               --------------  ----- 
 0   Finland$7.632$7.695$7.569$2.595$1.305$1.592$0.874$0.681$0.192$0.393  155 non-null    object
dtypes: object(1)
memory usage: 1.3+ KB


Finland$7.632$7.695$7.569$2.595$1.305$1.592$0.874$0.681$0.192$0.393
0  Norway$7.594$7.657$7.530$2.383$1.456$1.582$0.8...                 
1  Denmark$7.555$7.623$7.487$2.370$1.351$1.590$0....                 
2  Iceland$7.495$7.593$7.398$2.426$1.343$1.644$0....                 
3  Switzerland$7.487$7.570$7.405$2.320$1.420$1.54...                 
4  Netherlands$7.441$7.498$7.384$2.448$1.361$1.48...

In [17]:
df4 = pd.read_csv('data4.txt', sep="$", header = None)
df4.columns = df3.columns
df4.info()
df4.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     156 non-null    object 
 1   Happiness score                             156 non-null    float64
 2   Whisker-high                                156 non-null    float64
 3   Whisker-low                                 156 non-null    float64
 4   Dystopia (1.92) + residual                  156 non-null    float64
 5   Explained by: GDP per capita                156 non-null    float64
 6   Explained by: Social support                156 non-null    float64
 7   Explained by: Healthy life expectancy       156 non-null    float64
 8   Explained by: Freedom to make life choices  156 non-null    float64
 9   Explained by: Generosity                    156 non-null    float64
 10  Explained by: 

Country  Happiness score  Whisker-high  Whisker-low  \
0      Finland            7.632         7.695        7.569   
1       Norway            7.594         7.657        7.530   
2      Denmark            7.555         7.623        7.487   
3      Iceland            7.495         7.593        7.398   
4  Switzerland            7.487         7.570        7.405   

   Dystopia (1.92) + residual  Explained by: GDP per capita  \
0                       2.595                         1.305   
1                       2.383                         1.456   
2                       2.370                         1.351   
3                       2.426                         1.343   
4                       2.320                         1.420   

   Explained by: Social support  Explained by: Healthy life expectancy  \
0                         1.592                                  0.874   
1                         1.582                                  0.861   
2                         1.590                                  0.868   
3                         1.644                                  0.914   
4                         1.549                                  0.927   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                       0.681                     0.192   
1                                       0.686                     0.286   
2                                       0.683                     0.284   
3                                       0.677                     0.353   
4                                       0.660                     0.256   

   Explained by: Perceptions of corruption  
0                                    0.393  
1                                    0.340  
2                                    0.408  
3                                    0.138  
4                                    0.357

In [18]:
df4.describe()

Happiness score  Whisker-high  Whisker-low  Dystopia (1.92) + residual  \
count       156.000000    156.000000   156.000000                  156.000000   
mean          5.375878      5.478654     5.273090                    1.922891   
std           1.119507      1.107522     1.132544                    0.513760   
min           2.905000      3.074000     2.735000                    0.292000   
25%           4.453750      4.589750     4.344750                    1.654250   
50%           5.378000      5.477500     5.284500                    1.908500   
75%           6.168500      6.260000     6.051250                    2.270250   
max           7.632000      7.695000     7.569000                    2.961000   

       Explained by: GDP per capita  Explained by: Social support  \
count                    156.000000                    156.000000   
mean                       0.887417                      1.216571   
std                        0.382573                      0.300380   
min                        0.000000                      0.000000   
25%                        0.616250                      1.076750   
50%                        0.949500                      1.261500   
75%                        1.197750                      1.463000   
max                        1.649000                      1.644000   

       Explained by: Healthy life expectancy  \
count                             156.000000   
mean                                0.598026   
std                                 0.247925   
min                                 0.000000   
25%                                 0.422250   
50%                                 0.644000   
75%                                 0.777250   
max                                 1.030000   

       Explained by: Freedom to make life choices  Explained by: Generosity  \
count                                  156.000000                156.000000   
mean                                     0.456981                  0.181571   
std                                      0.162750                  0.098805   
min                                      0.000000                  0.000000   
25%                                      0.358250                  0.109500   
50%                                      0.494000                  0.174000   
75%                                      0.580000                  0.242250   
max                                      0.724000                  0.598000   

       Explained by: Perceptions of corruption  
count                               156.000000  
mean                                  0.112474  
std                                   0.096362  
min                                   0.000000  
25%                                   0.051000  
50%                                   0.082000  
75%                                   0.139000  
max                                   0.457000

This looked like a similar issue to the last one, however there were no column headers so I simply borrowed them from the previous dataset.

## Problem 5
Load `data5.csv`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

## Answer 5

In [19]:
df5 = pd.read_csv('data5.csv')
df5.info()
df5.tail(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158 entries, 0 to 157
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     158 non-null    object 
 1   Happiness score                             156 non-null    float64
 2   Whisker-high                                156 non-null    float64
 3   Whisker-low                                 156 non-null    float64
 4   Dystopia (1.92) + residual                  156 non-null    float64
 5   Explained by: GDP per capita                156 non-null    float64
 6   Explained by: Social support                156 non-null    float64
 7   Explained by: Healthy life expectancy       156 non-null    float64
 8   Explained by: Freedom to make life choices  156 non-null    float64
 9   Explained by: Generosity                    156 non-null    float64
 10  Explained by: 

Country  Happiness score  \
148                                            Liberia            3.495   
149                                              Syria            3.462   
150                                             Rwanda            3.408   
151                                              Yemen            3.355   
152                                           Tanzania            3.303   
153                                        South Sudan            3.254   
154                           Central African Republic            3.083   
155                                            Burundi            2.905   
156  Source: The World Happiness Report (2018), The...              NaN   
157          URL: http://worldhappiness.report/ed/2018              NaN   

     Whisker-high  Whisker-low  Dystopia (1.92) + residual  \
148         3.623        3.368                       1.640   
149         3.664        3.260                       1.244   
150         3.500        3.317                       0.500   
151         3.448        3.262                       1.106   
152         3.414        3.193                       0.628   
153         3.385        3.123                       1.691   
154         3.227        2.939                       2.487   
155         3.074        2.735                       1.752   
156           NaN          NaN                         NaN   
157           NaN          NaN                         NaN   

     Explained by: GDP per capita  Explained by: Social support  \
148                         0.076                         0.858   
149                         0.689                         0.382   
150                         0.332                         0.896   
151                         0.442                         1.073   
152                         0.455                         0.991   
153                         0.337                         0.608   
154                         0.024                         0.000   
155                         0.091                         0.627   
156                           NaN                           NaN   
157                           NaN                           NaN   

     Explained by: Healthy life expectancy  \
148                                  0.267   
149                                  0.539   
150                                  0.400   
151                                  0.343   
152                                  0.381   
153                                  0.177   
154                                  0.010   
155                                  0.145   
156                                    NaN   
157                                    NaN   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
148                                       0.419                     0.206   
149                                       0.088                     0.376   
150                                       0.636                     0.200   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   
156                                         NaN                       NaN   
157                                         NaN                       NaN   

     Explained by: Perceptions of corruption  
148                                    0.030  
149                                    0.144  
150                                    0.444  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076  
156                   

In [20]:
df5 = pd.read_csv('data5.csv', skipfooter = 2, engine='python')
df5.info()
df5.tail(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     156 non-null    object 
 1   Happiness score                             156 non-null    float64
 2   Whisker-high                                156 non-null    float64
 3   Whisker-low                                 156 non-null    float64
 4   Dystopia (1.92) + residual                  156 non-null    float64
 5   Explained by: GDP per capita                156 non-null    float64
 6   Explained by: Social support                156 non-null    float64
 7   Explained by: Healthy life expectancy       156 non-null    float64
 8   Explained by: Freedom to make life choices  156 non-null    float64
 9   Explained by: Generosity                    156 non-null    float64
 10  Explained by: 

Country  Happiness score  Whisker-high  Whisker-low  \
146                    Malawi            3.587         3.691        3.482   
147                     Haiti            3.582         3.713        3.451   
148                   Liberia            3.495         3.623        3.368   
149                     Syria            3.462         3.664        3.260   
150                    Rwanda            3.408         3.500        3.317   
151                     Yemen            3.355         3.448        3.262   
152                  Tanzania            3.303         3.414        3.193   
153               South Sudan            3.254         3.385        3.123   
154  Central African Republic            3.083         3.227        2.939   
155                   Burundi            2.905         3.074        2.735   

     Dystopia (1.92) + residual  Explained by: GDP per capita  \
146                       1.732                         0.186   
147                       1.744                         0.315   
148                       1.640                         0.076   
149                       1.244                         0.689   
150                       0.500                         0.332   
151                       1.106                         0.442   
152                       0.628                         0.455   
153                       1.691                         0.337   
154                       2.487                         0.024   
155                       1.752                         0.091   

     Explained by: Social support  Explained by: Healthy life expectancy  \
146                         0.541                                  0.306   
147                         0.714                                  0.289   
148                         0.858                                  0.267   
149                         0.382                                  0.539   
150                         0.896                                  0.400   
151                         1.073                                  0.343   
152                         0.991                                  0.381   
153                         0.608                                  0.177   
154                         0.000                                  0.010   
155                         0.627                                  0.145   

     Explained by: Freedom to make life choices  Explained by: Generosity  \
146                                       0.531                     0.210   
147                                       0.025                     0.392   
148                                       0.419                     0.206   
149                                       0.088                     0.376   
150                                       0.636                     0.200   
151                                       0.244                     0.083   
152                                       0.481                     0.270   
153                                       0.112                     0.224   
154                                       0.305                     0.218   
155                                       0.065                     0.149   

     Explained by: Perceptions of corruption  
146                                    0.080  
147                                    0.104  
148                                    0.030  
149                                    0.144  
150                                    0.444  
151                                    0.064  
152                                    0.097  
153                                    0.106  
154                                    0.038  
155                                    0.076

In [21]:
df5.describe()

Happiness score  Whisker-high  Whisker-low  Dystopia (1.92) + residual  \
count       156.000000    156.000000   156.000000                  156.000000   
mean          5.375878      5.478654     5.273090                    1.922891   
std           1.119507      1.107522     1.132544                    0.513760   
min           2.905000      3.074000     2.735000                    0.292000   
25%           4.453750      4.589750     4.344750                    1.654250   
50%           5.378000      5.477500     5.284500                    1.908500   
75%           6.168500      6.260000     6.051250                    2.270250   
max           7.632000      7.695000     7.569000                    2.961000   

       Explained by: GDP per capita  Explained by: Social support  \
count                    156.000000                    156.000000   
mean                       0.887417                      1.216571   
std                        0.382573                      0.300380   
min                        0.000000                      0.000000   
25%                        0.616250                      1.076750   
50%                        0.949500                      1.261500   
75%                        1.197750                      1.463000   
max                        1.649000                      1.644000   

       Explained by: Healthy life expectancy  \
count                             156.000000   
mean                                0.598026   
std                                 0.247925   
min                                 0.000000   
25%                                 0.422250   
50%                                 0.644000   
75%                                 0.777250   
max                                 1.030000   

       Explained by: Freedom to make life choices  Explained by: Generosity  \
count                                  156.000000                156.000000   
mean                                     0.456981                  0.181571   
std                                      0.162750                  0.098805   
min                                      0.000000                  0.000000   
25%                                      0.358250                  0.109500   
50%                                      0.494000                  0.174000   
75%                                      0.580000                  0.242250   
max                                      0.724000                  0.598000   

       Explained by: Perceptions of corruption  
count                               156.000000  
mean                                  0.112474  
std                                   0.096362  
min                                   0.000000  
25%                                   0.051000  
50%                                   0.082000  
75%                                   0.139000  
max                                   0.457000

Noticed that the null values weren't in the head but the tail this time. Easy fix with the `skipfooter` parameter.

## Problem 6
Load `data6.dat`. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (1 point)

## Answer 6

In [22]:
df6 = pd.read_csv('data6.dat')
df6.info()
df6.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     156 non-null    object 
 1   Happiness score                             156 non-null    float64
 2   Whisker-high                                156 non-null    float64
 3   Whisker-low                                 156 non-null    float64
 4   Dystopia (1.92) + residual                  156 non-null    float64
 5   Explained by: GDP per capita                156 non-null    float64
 6   Explained by: Social support                156 non-null    float64
 7   Explained by: Healthy life expectancy       156 non-null    float64
 8   Explained by: Freedom to make life choices  156 non-null    float64
 9   Explained by: Generosity                    156 non-null    float64
 10  Explained by: 

Happiness score  Whisker-high  Whisker-low  Dystopia (1.92) + residual  \
count       156.000000    156.000000   156.000000                  156.000000   
mean         94.623205    139.241705   132.737026                   72.229006   
std         284.884460    340.186886   333.268246                  256.083761   
min           2.905000      3.074000     2.939000                    0.292000   
25%           4.590500      4.710750     4.583750                    1.673000   
50%           5.672000      5.702500     5.654000                    1.977000   
75%           6.449750      6.854750     6.655250                    2.330000   
max         999.000000    999.000000   999.000000                  999.000000   

       Explained by: GDP per capita  Explained by: Social support  \
count                    156.000000                    156.000000   
mean                     122.463968                    141.938353   
std                      327.478649                    348.391531   
min                        0.000000                      0.000000   
25%                        0.689000                      1.149500   
50%                        1.016000                      1.333000   
75%                        1.293250                      1.518250   
max                      999.000000                    999.000000   

       Explained by: Healthy life expectancy  \
count                             156.000000   
mean                               90.191628   
std                               286.278522   
min                                 0.000000   
25%                                 0.439000   
50%                                 0.673000   
75%                                 0.858000   
max                               999.000000   

       Explained by: Freedom to make life choices  Explained by: Generosity  \
count                                  156.000000                156.000000   
mean                                   102.870276                 70.611955   
std                                    303.922857                256.530118   
min                                      0.000000                  0.000000   
25%                                      0.372000                  0.115500   
50%                                      0.517500                  0.182000   
75%                                      0.625250                  0.261250   
max                                    999.000000                999.000000   

       Explained by: Perceptions of corruption  
count                               156.000000  
mean                                 83.353615  
std                                 276.966930  
min                                   0.000000  
25%                                   0.052000  
50%                                   0.093000  
75%                                   0.171000  
max                                 999.000000

In [23]:
df6.head(10)

Country  Happiness score  Whisker-high  Whisker-low  \
0      Finland            7.632         7.695        7.569   
1       Norway            7.594         7.657        7.530   
2      Denmark            7.555         7.623        7.487   
3      Iceland            7.495         7.593      999.000   
4  Switzerland            7.487         7.570        7.405   
5  Netherlands            7.441         7.498        7.384   
6       Canada            7.328         7.421        7.236   
7  New Zealand            7.324         7.401        7.247   
8       Sweden            7.314         7.388        7.241   
9    Australia            7.272       999.000        7.186   

   Dystopia (1.92) + residual  Explained by: GDP per capita  \
0                       2.595                       999.000   
1                     999.000                       999.000   
2                       2.370                         1.351   
3                       2.426                         1.343   
4                       2.320                         1.420   
5                       2.448                         1.361   
6                       2.306                         1.330   
7                     999.000                         1.268   
8                       2.217                         1.355   
9                       2.139                         1.340   

   Explained by: Social support  Explained by: Healthy life expectancy  \
0                       999.000                                999.000   
1                         1.582                                999.000   
2                         1.590                                999.000   
3                         1.644                                  0.914   
4                         1.549                                  0.927   
5                         1.488                                  0.878   
6                         1.532                                  0.896   
7                         1.601                                  0.876   
8                         1.501                                  0.913   
9                         1.573                                  0.910   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                       0.681                     0.192   
1                                       0.686                     0.286   
2                                       0.683                     0.284   
3                                       0.677                     0.353   
4                                       0.660                     0.256   
5                                       0.638                     0.333   
6                                       0.653                     0.321   
7                                       0.669                     0.365   
8                                       0.659                   999.000   
9                                       0.647                     0.361   

   Explained by: Perceptions of corruption  
0                                    0.393  
1                                    0.340  
2                                    0.408  
3                                  999.000  
4                                    0.357  
5                                  999.000  
6                                    0.291  
7                                    0.389  
8                                    0.383  
9                                    0.302

There are some erroneous values of 999 that need to be addressed.

## Problem 7
Load `data7.xlsx`, which is an Excel file. Keep only the sheet named “Data”. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (2 points)

## Answer 7

In [24]:
df7 = pd.read_excel('data7.xlsx', sheet_name='Data')
df7.info()
df7.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     156 non-null    object 
 1   Happiness score                             156 non-null    float64
 2   Whisker-high                                156 non-null    float64
 3   Whisker-low                                 156 non-null    float64
 4   Dystopia (1.92) + residual                  156 non-null    float64
 5   Explained by: GDP per capita                156 non-null    float64
 6   Explained by: Social support                156 non-null    float64
 7   Explained by: Healthy life expectancy       156 non-null    float64
 8   Explained by: Freedom to make life choices  156 non-null    float64
 9   Explained by: Generosity                    156 non-null    float64
 10  Explained by: 

Country  Happiness score  Whisker-high  Whisker-low  \
0      Finland            7.632         7.695        7.569   
1       Norway            7.594         7.657        7.530   
2      Denmark            7.555         7.623        7.487   
3      Iceland            7.495         7.593        7.398   
4  Switzerland            7.487         7.570        7.405   
5  Netherlands            7.441         7.498        7.384   
6       Canada            7.328         7.421        7.236   
7  New Zealand            7.324         7.401        7.247   
8       Sweden            7.314         7.388        7.241   
9    Australia            7.272         7.358        7.186   

   Dystopia (1.92) + residual  Explained by: GDP per capita  \
0                       2.595                         1.305   
1                       2.383                         1.456   
2                       2.370                         1.351   
3                       2.426                         1.343   
4                       2.320                         1.420   
5                       2.448                         1.361   
6                       2.306                         1.330   
7                       2.156                         1.268   
8                       2.217                         1.355   
9                       2.139                         1.340   

   Explained by: Social support  Explained by: Healthy life expectancy  \
0                         1.592                                  0.874   
1                         1.582                                  0.861   
2                         1.590                                  0.868   
3                         1.644                                  0.914   
4                         1.549                                  0.927   
5                         1.488                                  0.878   
6                         1.532                                  0.896   
7                         1.601                                  0.876   
8                         1.501                                  0.913   
9                         1.573                                  0.910   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                       0.681                     0.192   
1                                       0.686                     0.286   
2                                       0.683                     0.284   
3                                       0.677                     0.353   
4                                       0.660                     0.256   
5                                       0.638                     0.333   
6                                       0.653                     0.321   
7                                       0.669                     0.365   
8                                       0.659                     0.285   
9                                       0.647                     0.361   

   Explained by: Perceptions of corruption  
0                                    0.393  
1                                    0.340  
2                                    0.408  
3                                    0.138  
4                                    0.357  
5                                    0.295  
6                                    0.291  
7                                    0.389  
8                                    0.383  
9                                    0.302

In [25]:
df7.describe()

Happiness score  Whisker-high  Whisker-low  Dystopia (1.92) + residual  \
count       156.000000    156.000000   156.000000                  156.000000   
mean          5.375878      5.478654     5.273090                    1.922891   
std           1.119507      1.107522     1.132544                    0.513760   
min           2.905000      3.074000     2.735000                    0.292000   
25%           4.453750      4.589750     4.344750                    1.654250   
50%           5.378000      5.477500     5.284500                    1.908500   
75%           6.168500      6.260000     6.051250                    2.270250   
max           7.632000      7.695000     7.569000                    2.961000   

       Explained by: GDP per capita  Explained by: Social support  \
count                    156.000000                    156.000000   
mean                       0.887417                      1.216571   
std                        0.382573                      0.300380   
min                        0.000000                      0.000000   
25%                        0.616250                      1.076750   
50%                        0.949500                      1.261500   
75%                        1.197750                      1.463000   
max                        1.649000                      1.644000   

       Explained by: Healthy life expectancy  \
count                             156.000000   
mean                                0.598026   
std                                 0.247925   
min                                 0.000000   
25%                                 0.422250   
50%                                 0.644000   
75%                                 0.777250   
max                                 1.030000   

       Explained by: Freedom to make life choices  Explained by: Generosity  \
count                                  156.000000                156.000000   
mean                                     0.456981                  0.181571   
std                                      0.162750                  0.098805   
min                                      0.000000                  0.000000   
25%                                      0.358250                  0.109500   
50%                                      0.494000                  0.174000   
75%                                      0.580000                  0.242250   
max                                      0.724000                  0.598000   

       Explained by: Perceptions of corruption  
count                               156.000000  
mean                                  0.112474  
std                                   0.096362  
min                                   0.000000  
25%                                   0.051000  
50%                                   0.082000  
75%                                   0.139000  
max                                   0.457000

It seems like all that was needed as passing the parameter of `sheet_name = 'Data'`.

## Problem 8
Load `data8.dta`, which is a Stata 13 file. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (2 points)

## Answer 8

In [26]:
df8 = pd.read_stata("data8.dta")
df8.info()
df8.head(10)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   country                           156 non-null    object 
 1   happinessscore                    156 non-null    float32
 2   whiskerhigh                       156 non-null    float32
 3   whiskerlow                        156 non-null    float32
 4   dystopia192residual               156 non-null    float32
 5   explainedbygdppercapita           156 non-null    float32
 6   explainedbysocialsupport          156 non-null    float32
 7   explainedbyhealthylifeexpectancy  156 non-null    float32
 8   explainedbyfreedomtomakelifechoi  156 non-null    float32
 9   explainedbygenerosity             156 non-null    float32
 10  explainedbyperceptionsofcorrupti  156 non-null    float32
dtypes: float32(10), object(1)
memory usage: 8.5+ KB


country  happinessscore  whiskerhigh  whiskerlow  dystopia192residual  \
0      Finland           7.632        7.695       7.569                2.595   
1       Norway           7.594        7.657       7.530                2.383   
2      Denmark           7.555        7.623       7.487                2.370   
3      Iceland           7.495        7.593       7.398                2.426   
4  Switzerland           7.487        7.570       7.405                2.320   
5  Netherlands           7.441        7.498       7.384                2.448   
6       Canada           7.328        7.421       7.236                2.306   
7  New Zealand           7.324        7.401       7.247                2.156   
8       Sweden           7.314        7.388       7.241                2.217   
9    Australia           7.272        7.358       7.186                2.139   

   explainedbygdppercapita  explainedbysocialsupport  \
0                    1.305                     1.592   
1                    1.456                     1.582   
2                    1.351                     1.590   
3                    1.343                     1.644   
4                    1.420                     1.549   
5                    1.361                     1.488   
6                    1.330                     1.532   
7                    1.268                     1.601   
8                    1.355                     1.501   
9                    1.340                     1.573   

   explainedbyhealthylifeexpectancy  explainedbyfreedomtomakelifechoi  \
0                             0.874                             0.681   
1                             0.861                             0.686   
2                             0.868                             0.683   
3                             0.914                             0.677   
4                             0.927                             0.660   
5                             0.878                             0.638   
6                             0.896                             0.653   
7                             0.876                             0.669   
8                             0.913                             0.659   
9                             0.910                             0.647   

   explainedbygenerosity  explainedbyperceptionsofcorrupti  
0                  0.192                             0.393  
1                  0.286                             0.340  
2                  0.284                             0.408  
3                  0.353                             0.138  
4                  0.256                             0.357  
5                  0.333                             0.295  
6                  0.321                             0.291  
7                  0.365                             0.389  
8                  0.285                             0.383  
9                  0.361                             0.302

In [27]:
df8.describe()

happinessscore  whiskerhigh  whiskerlow  dystopia192residual  \
count      156.000000   156.000000  156.000000           156.000000   
mean         5.375877     5.478654    5.273088             1.922891   
std          1.119507     1.107522    1.132544             0.513760   
min          2.905000     3.074000    2.735000             0.292000   
25%          4.453750     4.589750    4.344750             1.654250   
50%          5.378000     5.477500    5.284500             1.908500   
75%          6.168500     6.260000    6.051250             2.270250   
max          7.632000     7.695000    7.569000             2.961000   

       explainedbygdppercapita  explainedbysocialsupport  \
count               156.000000                 156.00000   
mean                  0.887417                   1.21657   
std                   0.382573                   0.30038   
min                   0.000000                   0.00000   
25%                   0.616250                   1.07675   
50%                   0.949500                   1.26150   
75%                   1.197750                   1.46300   
max                   1.649000                   1.64400   

       explainedbyhealthylifeexpectancy  explainedbyfreedomtomakelifechoi  \
count                        156.000000                        156.000000   
mean                           0.598026                          0.456981   
std                            0.247925                          0.162750   
min                            0.000000                          0.000000   
25%                            0.422250                          0.358250   
50%                            0.644000                          0.494000   
75%                            0.777250                          0.580000   
max                            1.030000                          0.724000   

       explainedbygenerosity  explainedbyperceptionsofcorrupti  
count             156.000000                        156.000000  
mean                0.181571                          0.112474  
std                 0.098805                          0.096362  
min                 0.000000                          0.000000  
25%                 0.109500                          0.051000  
50%                 0.174000                          0.082000  
75%                 0.242250                          0.139000  
max                 0.598000                          0.457000

All that was needed was using the `read_stata` function.

## Problem 9
Load `data9.sav`, which is an SPSS file. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (2 points)

## Answer 9

In [28]:
df9 = pd.read_spss("data9.sav")
df9.info()
df9.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   country      156 non-null    object 
 1   happiness    156 non-null    float64
 2   whiskerhigh  156 non-null    float64
 3   whiskerlow   156 non-null    float64
 4   dystopia     156 non-null    float64
 5   gdpPC        156 non-null    float64
 6   socsupport   156 non-null    float64
 7   lifeexp      156 non-null    float64
 8   lifechoice   156 non-null    float64
 9   generous     156 non-null    float64
 10  corrupt      156 non-null    float64
dtypes: float64(10), object(1)
memory usage: 13.5+ KB


country  happiness  whiskerhigh  whiskerlow  dystopia  gdpPC  \
0      Finland      7.632        7.695       7.569     2.595  1.305   
1       Norway      7.594        7.657       7.530     2.383  1.456   
2      Denmark      7.555        7.623       7.487     2.370  1.351   
3      Iceland      7.495        7.593       7.398     2.426  1.343   
4  Switzerland      7.487        7.570       7.405     2.320  1.420   
5  Netherlands      7.441        7.498       7.384     2.448  1.361   
6       Canada      7.328        7.421       7.236     2.306  1.330   
7  New Zealand      7.324        7.401       7.247     2.156  1.268   
8       Sweden      7.314        7.388       7.241     2.217  1.355   
9    Australia      7.272        7.358       7.186     2.139  1.340   

   socsupport  lifeexp  lifechoice  generous  corrupt  
0       1.592    0.874       0.681     0.192    0.393  
1       1.582    0.861       0.686     0.286    0.340  
2       1.590    0.868       0.683     0.284    0.408  
3       1.644    0.914       0.677     0.353    0.138  
4       1.549    0.927       0.660     0.256    0.357  
5       1.488    0.878       0.638     0.333    0.295  
6       1.532    0.896       0.653     0.321    0.291  
7       1.601    0.876       0.669     0.365    0.389  
8       1.501    0.913       0.659     0.285    0.383  
9       1.573    0.910       0.647     0.361    0.302

In [29]:
df9.describe()

happiness  whiskerhigh  whiskerlow    dystopia       gdpPC  \
count  156.000000   156.000000  156.000000  156.000000  156.000000   
mean     5.375878     5.478654    5.273090    1.922891    0.887417   
std      1.119507     1.107522    1.132544    0.513760    0.382573   
min      2.905000     3.074000    2.735000    0.292000    0.000000   
25%      4.453750     4.589750    4.344750    1.654250    0.616250   
50%      5.378000     5.477500    5.284500    1.908500    0.949500   
75%      6.168500     6.260000    6.051250    2.270250    1.197750   
max      7.632000     7.695000    7.569000    2.961000    1.649000   

       socsupport     lifeexp  lifechoice    generous     corrupt  
count  156.000000  156.000000  156.000000  156.000000  156.000000  
mean     1.216571    0.598026    0.456981    0.181571    0.112474  
std      0.300380    0.247925    0.162750    0.098805    0.096362  
min      0.000000    0.000000    0.000000    0.000000    0.000000  
25%      1.076750    0.422250    0.358250    0.109500    0.051000  
50%      1.261500    0.644000    0.494000    0.174000    0.082000  
75%      1.463000    0.777250    0.580000    0.242250    0.139000  
max      1.644000    1.030000    0.724000    0.598000    0.457000

Similar to the last problem, I just used `read_spss` this time instead.

## Problem 10
Load `data10.xpt`, which is a SAS file. Use the tools we discussed in class to decide whether the data file loaded correctly, and include that code in your lab report. In one or two sentences, describe how you decided on the right combination of parameters needed to load the data. (If some of the country names display as `b'Finland'`, don't worry about that.) (2 points)

## Answer 10

In [30]:
df10 = pd.read_sas("data10.xpt", encoding="utf-8")
df10.info()
df10.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   COUNTRY   156 non-null    object 
 1   HAPPINES  156 non-null    float64
 2   WHISKERH  156 non-null    float64
 3   WHISKERL  156 non-null    float64
 4   DYSTOPIA  156 non-null    float64
 5   EXPLAINE  156 non-null    float64
 6   EXPLAIN2  156 non-null    float64
 7   EXPLAIN3  156 non-null    float64
 8   EXPLAIN4  156 non-null    float64
 9   EXPLAIN5  156 non-null    float64
 10  EXPLAIN6  156 non-null    float64
dtypes: float64(10), object(1)
memory usage: 13.5+ KB


COUNTRY  HAPPINES  WHISKERH  WHISKERL  DYSTOPIA  EXPLAINE  EXPLAIN2  \
0      Finland     7.632     7.695     7.569     2.595     1.305     1.592   
1       Norway     7.594     7.657     7.530     2.383     1.456     1.582   
2      Denmark     7.555     7.623     7.487     2.370     1.351     1.590   
3      Iceland     7.495     7.593     7.398     2.426     1.343     1.644   
4  Switzerland     7.487     7.570     7.405     2.320     1.420     1.549   
5  Netherlands     7.441     7.498     7.384     2.448     1.361     1.488   
6       Canada     7.328     7.421     7.236     2.306     1.330     1.532   
7  New Zealand     7.324     7.401     7.247     2.156     1.268     1.601   
8       Sweden     7.314     7.388     7.241     2.217     1.355     1.501   
9    Australia     7.272     7.358     7.186     2.139     1.340     1.573   

   EXPLAIN3  EXPLAIN4  EXPLAIN5  EXPLAIN6  
0     0.874     0.681     0.192     0.393  
1     0.861     0.686     0.286     0.340  
2     0.868     0.683     0.284     0.408  
3     0.914     0.677     0.353     0.138  
4     0.927     0.660     0.256     0.357  
5     0.878     0.638     0.333     0.295  
6     0.896     0.653     0.321     0.291  
7     0.876     0.669     0.365     0.389  
8     0.913     0.659     0.285     0.383  
9     0.910     0.647     0.361     0.302

In [31]:
df10.describe()

HAPPINES    WHISKERH    WHISKERL    DYSTOPIA      EXPLAINE  \
count  156.000000  156.000000  156.000000  156.000000  1.560000e+02   
mean     5.375878    5.478654    5.273090    1.922891  8.874167e-01   
std      1.119507    1.107522    1.132544    0.513760  3.825727e-01   
min      2.905000    3.074000    2.735000    0.292000  5.397605e-79   
25%      4.453750    4.589750    4.344750    1.654250  6.162500e-01   
50%      5.378000    5.477500    5.284500    1.908500  9.495000e-01   
75%      6.168500    6.260000    6.051250    2.270250  1.197750e+00   
max      7.632000    7.695000    7.569000    2.961000  1.649000e+00   

           EXPLAIN2      EXPLAIN3      EXPLAIN4      EXPLAIN5      EXPLAIN6  
count  1.560000e+02  1.560000e+02  1.560000e+02  1.560000e+02  1.560000e+02  
mean   1.216571e+00  5.980256e-01  4.569808e-01  1.815705e-01  1.124744e-01  
std    3.003804e-01  2.479247e-01  1.627499e-01  9.880460e-02  9.636232e-02  
min    5.397605e-79  5.397605e-79  5.397605e-79  5.397605e-79  5.397605e-79  
25%    1.076750e+00  4.222500e-01  3.582500e-01  1.095000e-01  5.100000e-02  
50%    1.261500e+00  6.440000e-01  4.940000e-01  1.740000e-01  8.200000e-02  
75%    1.463000e+00  7.772500e-01  5.800000e-01  2.422500e-01  1.390000e-01  
max    1.644000e+00  1.030000e+00  7.240000e-01  5.980000e-01  4.570000e-01

I worried about countries displaying as `b'Finland'` so I added the encoding parameter. Other than that though everything looked to be good.

## Problem 11
Please load the `data11.txt` file, which is a fixed width file. The columns are defined as follows:

| Variable                                   | Width | Start | End |
|--------------------------------------------|-------|-------|-----|
| Country                                    | 24    | 1     | 24  |
| Happiness score                            | 5     | 25    | 29  |
| Whisker-high                               | 5     | 30    | 34  |
| Whisker-low                                | 5     | 35    | 39  |
| Dystopia (1.92) + residual                 | 5     | 40    | 44  |
| Explained by: GDP per capita               | 5     | 45    | 49  |
| Explained by: Social support               | 5     | 50    | 54  |
| Explained by: Healthy life expectancy      | 5     | 55    | 59  |
| Explained by: Freedom to make life choices | 5     | 60    | 64  |
| Explained by: Generosity                   | 5     | 65    | 69  |
| Explained by: Perceptions of corruption    | 5     | 70    | 74  |

Then save the this loaded data frame as a CSV file on your local machine. Be sure to use a unique filename so as not to overwrite any existing files. (5 points)

## Answer 11

In [32]:
headings = ['Country', 'Happiness score', 'Whisker-high', 'Whisker-low', 'Dystopia (1.92) + residual', 'Explained by: GDP per capita',
                    'Explained by: Social support', 'Explained by: Healthy life expectancy', 'Explained by: Freedom to make life choices',
            'Explained by: Generosity', 'Explained by: Perceptions of corruption']

colspecs = [(0, 23), (24, 29), (29, 34), (34, 39), (39, 44), (44, 49),
                    (49, 54), (54, 59), (59, 64), (64, 69), (69, 74)]

df11 = pd.read_fwf("data11.txt", names=headings,
                           header=None, colspecs=colspecs)
df11.info()
df11.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     156 non-null    object 
 1   Happiness score                             156 non-null    float64
 2   Whisker-high                                156 non-null    float64
 3   Whisker-low                                 156 non-null    float64
 4   Dystopia (1.92) + residual                  156 non-null    float64
 5   Explained by: GDP per capita                156 non-null    float64
 6   Explained by: Social support                156 non-null    float64
 7   Explained by: Healthy life expectancy       156 non-null    float64
 8   Explained by: Freedom to make life choices  156 non-null    float64
 9   Explained by: Generosity                    156 non-null    float64
 10  Explained by: 

Country  Happiness score  Whisker-high  Whisker-low  \
0      Finland            7.632         7.695        7.569   
1       Norway            7.594         7.657        7.530   
2      Denmark            7.555         7.623        7.487   
3      Iceland            7.495         7.593        7.398   
4  Switzerland            7.487         7.570        7.405   
5  Netherlands            7.441         7.498        7.384   
6       Canada            7.328         7.421        7.236   
7  New Zealand            7.324         7.401        7.247   
8       Sweden            7.314         7.388        7.241   
9    Australia            7.272         7.358        7.186   

   Dystopia (1.92) + residual  Explained by: GDP per capita  \
0                       2.595                         1.305   
1                       2.383                         1.456   
2                       2.370                         1.351   
3                       2.426                         1.343   
4                       2.320                         1.420   
5                       2.448                         1.361   
6                       2.306                         1.330   
7                       2.156                         1.268   
8                       2.217                         1.355   
9                       2.139                         1.340   

   Explained by: Social support  Explained by: Healthy life expectancy  \
0                         1.592                                  0.874   
1                         1.582                                  0.861   
2                         1.590                                  0.868   
3                         1.644                                  0.914   
4                         1.549                                  0.927   
5                         1.488                                  0.878   
6                         1.532                                  0.896   
7                         1.601                                  0.876   
8                         1.501                                  0.913   
9                         1.573                                  0.910   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                       0.681                     0.192   
1                                       0.686                     0.286   
2                                       0.683                     0.284   
3                                       0.677                     0.353   
4                                       0.660                     0.256   
5                                       0.638                     0.333   
6                                       0.653                     0.321   
7                                       0.669                     0.365   
8                                       0.659                     0.285   
9                                       0.647                     0.361   

   Explained by: Perceptions of corruption  
0                                    0.393  
1                                    0.340  
2                                    0.408  
3                                    0.138  
4                                    0.357  
5                                    0.295  
6                                    0.291  
7                                    0.389  
8                                    0.383  
9                                    0.302

In [33]:
df11.describe()

Happiness score  Whisker-high  Whisker-low  Dystopia (1.92) + residual  \
count       156.000000    156.000000   156.000000                  156.000000   
mean          5.375878      5.478654     5.273090                    1.922891   
std           1.119507      1.107522     1.132544                    0.513760   
min           2.905000      3.074000     2.735000                    0.292000   
25%           4.453750      4.589750     4.344750                    1.654250   
50%           5.378000      5.477500     5.284500                    1.908500   
75%           6.168500      6.260000     6.051250                    2.270250   
max           7.632000      7.695000     7.569000                    2.961000   

       Explained by: GDP per capita  Explained by: Social support  \
count                    156.000000                    156.000000   
mean                       0.887417                      1.216571   
std                        0.382573                      0.300380   
min                        0.000000                      0.000000   
25%                        0.616250                      1.076750   
50%                        0.949500                      1.261500   
75%                        1.197750                      1.463000   
max                        1.649000                      1.644000   

       Explained by: Healthy life expectancy  \
count                             156.000000   
mean                                0.598026   
std                                 0.247925   
min                                 0.000000   
25%                                 0.422250   
50%                                 0.644000   
75%                                 0.777250   
max                                 1.030000   

       Explained by: Freedom to make life choices  Explained by: Generosity  \
count                                  156.000000                156.000000   
mean                                     0.456981                  0.181571   
std                                      0.162750                  0.098805   
min                                      0.000000                  0.000000   
25%                                      0.358250                  0.109500   
50%                                      0.494000                  0.174000   
75%                                      0.580000                  0.242250   
max                                      0.724000                  0.598000   

       Explained by: Perceptions of corruption  
count                               156.000000  
mean                                  0.112474  
std                                   0.096362  
min                                   0.000000  
25%                                   0.051000  
50%                                   0.082000  
75%                                   0.139000  
max                                   0.457000

In [34]:
df11.to_csv('data11.csv')  

First step was to make the headings and column spacings as specified, then we could read in the fixed width file using the `read_fwf` function. Then I verified the contents and saved to a .csv file.